core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -0,0 +1,398 @@
+ * Producers with no specific transactional id may talk to a random broker as their coordinators.
+ */
+class TransactionCoordinator(brokerId: Int,
+                             pidManager: ProducerIdManager,
+                             txnManager: TransactionStateManager,
+                             txnMarkerChannelManager: TransactionMarkerChannelManager,
+                             txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker],
+                             time: Time) extends Logging {
+  this.logIdent = "[Transaction Coordinator " + brokerId + "]: "
+
+  import TransactionCoordinator._
+
+  type InitPidCallback = InitPidResult => Unit
+  type TxnMetadataUpdateCallback = Errors => Unit
+  type EndTxnCallback = Errors => Unit
+
+  /* Active flag of the coordinator */
+  private val isActive = new AtomicBoolean(false)
+
+  def handleInitPid(transactionalId: String,
+                    transactionTimeoutMs: Int,
+                    responseCallback: InitPidCallback): Unit = {
+    if (transactionalId == null || transactionalId.isEmpty) {
+      // if the transactional id is not specified, then always blindly accept the request
+      // and return a new pid from the pid manager
+      val pid = pidManager.nextPid()
+      responseCallback(InitPidResult(pid, epoch = 0, Errors.NONE))
+    } else if (!txnManager.isCoordinatorFor(transactionalId)) {
+      // check if it is the assigned coordinator for the transactional id
+      responseCallback(initTransactionError(Errors.NOT_COORDINATOR))
+    } else if (txnManager.isCoordinatorLoadingInProgress(transactionalId)) {
+      responseCallback(initTransactionError(Errors.COORDINATOR_LOAD_IN_PROGRESS))
+    } else if (!txnManager.validateTransactionTimeoutMs(transactionTimeoutMs)) {
+      // check transactionTimeoutMs is not larger than the broker configured maximum allowed value
+      responseCallback(initTransactionError(Errors.INVALID_TRANSACTION_TIMEOUT))
+    } else {
+      // only try to get a new pid and update the cache if the transactional id is unknown
+      txnManager.getTransactionState(transactionalId) match {
+        case None =>
+          val pid = pidManager.nextPid()
+          val newMetadata: TransactionMetadata = new TransactionMetadata(pid = pid,
+            producerEpoch = 0,
+            txnTimeoutMs = transactionTimeoutMs,
+            state = Empty,
+            topicPartitions = collection.mutable.Set.empty[TopicPartition],
+            transactionStartTime = time.milliseconds(),
+            entryTimestamp = time.milliseconds())
+
+          val metadata = txnManager.addTransaction(transactionalId, newMetadata)
+
+          // there might be a concurrent thread that has just updated the mapping
+          // with the transactional id at the same time; in this case we will
+          // treat it as the metadata has existed and update it accordingly
```

Comment:
I think @junrao 's comment is that we did some checking on the txn metadata's state in `initPidWithExistingMetadata` whereas we did not do such checking before calling `appendMetadataToLog`. Have explained to him that it is because at line 129 we are assured that the metadata is just newly created and hence it's always `Ongoing`. Maybe the comment itself has been outdated after the addition of the `initPidWithExistingMetadata` logic.